@pblottiere : I've added a test that highlight the problem.
Here is the output of the test (before the proposed fix): 
```
574: ======================================================================
574: FAIL: test_fcgiRequestBody (__main__.QgsServerRequestTest)
574: Test request body
574: ----------------------------------------------------------------------
574: Traceback (most recent call last):
574:   File "/home/amorvan/dev/QGIS/tests/src/python/test_qgsserver_request.py", line 210, in test_fcgiRequestBody
574:     self.assertEqual(request.parameter('REQUEST_BODY'), '<Literal>+1</Literal>')
574: AssertionError: '<Literal> 1</Literal>' != '<Literal>+1</Literal>'
574: - <Literal> 1</Literal>
574: ?          ^
574: + <Literal>+1</Literal>
574: ?          ^
574: 
```

--- a/src/server/qgsrequesthandler.cpp
+++ b/src/server/qgsrequesthandler.cpp
@@ -252,7 +252,7 @@ void QgsRequestHandler::parseInput()
 
           mRequest.setParameter( attrName.toUpper(), attr.value() );
         }
-        mRequest.setParameter( QStringLiteral( "REQUEST_BODY" ), inputString );
+        mRequest.setParameter( QStringLiteral( "REQUEST_BODY" ), inputString.replace( '+', QStringLiteral( "%2B" ) ) );
       }
     }
   }

--- a/tests/src/python/test_qgsserver_request.py
+++ b/tests/src/python/test_qgsserver_request.py
@@ -194,6 +194,21 @@ class QgsServerRequestTest(QgsServerTestBase):
         _check_links(params)
         _check_links(params, 'POST')
 
+    def test_fcgiRequestBody(self):
+        """Test request body"""
+        data = '<Literal>+1</Literal>'
+        self._set_env({
+            'SERVER_NAME': 'www.myserver.com',
+            'SERVICE': 'WFS',
+            'REQUEST_BODY': data,
+            'CONTENT_LENGTH': str(len(data)),
+            'REQUEST_METHOD': 'POST',
+        })
+        request = QgsFcgiServerRequest()
+        response = QgsBufferServerResponse()
+        self.server.handleRequest(request, response)
+        self.assertEqual(request.parameter('REQUEST_BODY'), '<Literal>+1</Literal>')
+
     def test_add_parameters(self):
         request = QgsServerRequest()
         request.setParameter('FOOBAR', 'foobar')

